### ISSUE TITLE
feat: InPlacePodVerticalScaling support

### ISSUE THREAD

**Issue by jzhn** [kubernetes-sigs/karpenter#829]
### Description

**What problem are you trying to solve?**

Currently, Karpenter make scheduling and disruption decisions with the assumption that pod resource requests are immutable. Since Kubernetes 1.27, [InPlacePodVerticalScaling](https://kubernetes.io/blog/2023/05/12/in-place-pod-resize-alpha/) was introduced as a new alpha feature, It's targeting beta post 1.30. With InPlacePodVerticalScaling, pod resource requests and limits become mutable.

A common use case of InPlacePodVerticalScaling is to mitigate startup issues for heavy application like Java, that big resource requests is initially allocated at startup, when the Pod becomes ready, a controller then lowers the resource requests to free up resources on the node.

With current Karpenter implementation, it is possible to create a loop that
- pods run in idle state with low resource requests
- Karpenter determines node pool under utilized, evict and consolidation is triggered
- new pods start up with big resource requests, they become unscheduled
- new nodes are bought up to schedule new pods

Karpenter need to be updated to recognize the mutable resource requests, to prevent such loop. Due to the flexibility of current InPlacePodVerticalScaling, this might be a difficult task if Karpenter itself doesn't understand the resource request mutation strategy.

**How important is this feature to you?**

Many users have shown interest in InPlacePodVerticalScaling feature. https://github.com/aws/containers-roadmap/issues/512 might provide some datapoint. As a cluster autoscaler, Karpenter's awareness of InPlacePodVerticalScaling is critical so together the node usage efficiency can be further improved while keeping applications stable and performant.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jzhn** [comment]
@tzneal : I believe this issue is still relavant, Karpenter should at least provide some guidelines in its doc to mention the interactions between Karpenter consolidation and `InPlacePodVerticalScaling`. Do you mind if we reopen this one for tracking? (or if there have been new duplicates created please link it).

**Comment by tzneal** [comment]
Yes, sorry, this should not have gotten closed.

**Comment by jonathan-innis** [comment]
/assign @Bryce-Soghigian 

Pulling Bryce as the assignee to this issue since he was the assignee in the other (duplicate) issue that was tracking this

**Comment by vazkarvishal** [comment]
It will be good to know when this is planned to be implemented/tested, as we are also looking to use this feature

**Comment by Bryce-Soghigian** [comment]
Hi @vazkarvishal, It looks like InPlacePodVerticalScaling is planned to go beta in 1.33. 

[src](https://kubernetes.slack.com/archives/C09R1LV8S/p1741349380770089?thread_ts=1711569356.720629&cid=C09R1LV8S)

I'm thinking we should wait for it to land as beta and live for a bit before implementing support in karpenter. Maybe soonish, would be good to gauge other community interest in the feature. 

There is a kubernetes slack channel for the feature #sig-node-inplace-pod-resize for the feature.

**Comment by sftim** [comment]
Karpenter kind of does need to know if a node that previously had 75% free capacity is now 100% used, as it affects scheduling  / preemption.

IMO, Karpenter core can add support behind a feature gate or build-time opt in control, right now (even before the v1.33 release). Actual implementations might want to hold off longer before making even informal commitments to users.

**Comment by kfirsch** [comment]
With EKS 1.33 rolling out in few days and InPlacePodVerticalScaling enabled by default, this issue becomes more urgent. It would be great to know if there’s any roadmap or technical blockers for Karpenter to properly handle updated resource requests, especially since node utilization tracking and consolidation logic might break or misbehave without awareness of these changes. Any updates from maintainers would be appreciated.

**Comment by omri-shilton** [comment]
any update on this?

**Comment by ajaykumarmandapati** [comment]
Is there any news on when/if this will be implemented in Karpenter? 🤔

**Comment by SirNorlan** [comment]
any update here? This would be great to have across 100% of our customer base

**Comment by AlexandreDemailly** [comment]
Any news on this ?

**Comment by guruprasathselvaraj** [comment]
It would be greatly appreciated if this issue could be addressed as soon as possible. This feature plays a key role in supporting Karpenter’s core motto and overall mission—reducing costs and increasing overall cluster utilization.

**Comment by nathangeology** [comment]
Alright everyone, I see that there is persistent interest in this issue. In order to move forward on some set of changes related to this we need a couple of things. One, we need some use cases to test against. 

One that was mentioned in the initial description is a case where you need a lot of resources at start-up and then fewer later on once the application is up and running. For this use case, how long is this start-up period? What is the impact if the application experiences resource contention during startup? What is the magnitude of difference between the resources required at start-up vs when the application is stable? Which specific resources are we scaling (I assume CPU, but is that correct??). 

Are there other use cases we should consider here? 

Why is setting the requested resources to the 'stable' application demands and the limit to the burst-y application start-up level demand insufficient here? Is it because of resource contention on new nodes where all the pods are starting together? Are there other reasons this doesn't work for you here?

The next questions I have are more about Karpenter's potential interactions with some sort of feature around vertical pod autoscaling and need some setup. So take the use case above of high start-up application demand followed by a vertical pod request scale-in. Karpenter today would provision some nodes based on the initial pod request and wouldn't consider the eventual expected pod request size at all during placement. The pods are placed, they get through the initial start-up phase and then scale-in vertically to a smaller size. The nodes are all now under-utilized. 

First question: What is the expected behavior of Consolidation for ConsolidateWhenEmptyOrUnderutilized here?

 Right now, it would go and disrupt some of the nodes and which would then stop running pods and restart them on existing nodes. The pods would come in at the start-up, bigger pod size. So you'd be paying the start-up tax multiple times in row to try to pack the stable, smaller post start-up pods onto fewer, smaller, nodes over time. Plus, if you aren't careful you could create cycles where the pod gets places -> startup completes, pod scales down vertically -> node is now under-utilized -> node gets consolidated -> pod gets placed on new node..... on so on forever. Stability isn't assured with this setup, even if you tell consolidation about the size the pod would be if you had to replace it, you could still have longer stabilization periods and increased disruption costs overwhelm any savings here. 

How could we fix that? Well, we could try to slow down during provisioning and hold onto pods for a bit when we know some capacity is going to be freed by vertical scale-ins in the near future. You still have some coordination problems with consolidation there (it would be sad if you held a pod for a node that was going to get some capacity back and then consolidation got the node before you could place the pod there). This also directly trades-off of pod scaling responsiveness for packing efficiency and involves a whole slew of configuration to drive the behavior you want. 

This leads to my final question, what is the behavior you are looking for here? How do we keep the consolidation process from causing to much disruption and cancelling out the potential savings? What config setups do you imagine doing this in? Are we only talking consolidateWhenEmpty or DoNotConsolidate or static nodepools here (you can probably use this feature with those nodepool settings today!)? Are you envisioning this working with ConsolidateWhenEmptyOrUnderutilized? How do you envision this working, can you walk through some scenarios of how this should work without causing feedback loops or excessive disruption? 

Thanks! The team is busy and can always use help and contributions. I'm happy to work through this and help get it over the line once I understand more concretely what desired behavior is and have some ideas about how we can ensure it doesn't interact in unsafe ways with consolidation/provisioning.

**Comment by ajaykumarmandapati** [comment]
> Alright everyone, I see that there is persistent interest in this issue. In order to move forward on some set of changes related to this we need a couple of things. One, we need some use cases to test against.
> 
> One that was mentioned in the initial description is a case where you need a lot of resources at start-up and then fewer later on once the application is up and running. For this use case, how long is this start-up period? What is the impact if the application experiences resource contention during startup? What is the magnitude of difference between the resources required at start-up vs when the application is stable? Which specific resources are we scaling (I assume CPU, but is that correct??).
> 
> Are there other use cases we should consider here?
> 
> Why is setting the requested resources to the 'stable' application demands and the limit to the burst-y application start-up level demand insufficient here? Is it because of resource contention on new nodes where all the pods are starting together? Are there other reasons this doesn't work for you here?
> 
> The next questions I have are more about Karpenter's potential interactions with some sort of feature around vertical pod autoscaling and need some setup. So take the use case above of high start-up application demand followed by a vertical pod request scale-in. Karpenter today would provision some nodes based on the initial pod request and wouldn't consider the eventual expected pod request size at all during placement. The pods are placed, they get through the initial start-up phase and then scale-in vertically to a smaller size. The nodes are all now under-utilized.
> 
> First question: What is the expected behavior of Consolidation for ConsolidateWhenEmptyOrUnderutilized here?
> 
> Right now, it would go and disrupt some of the nodes and which would then stop running pods and restart them on existing nodes. The pods would come in at the start-up, bigger pod size. So you'd be paying the start-up tax multiple times in row to try to pack the stable, smaller post start-up pods onto fewer, smaller, nodes over time. Plus, if you aren't careful you could create cycles where the pod gets places -> startup completes, pod scales down vertically -> node is now under-utilized -> node gets consolidated -> pod gets placed on new node..... on so on forever. Stability isn't assured with this setup, even if you tell consolidation about the size the pod would be if you had to replace it, you could still have longer stabilization periods and increased disruption costs overwhelm any savings here.
> 
> How could we fix that? Well, we could try to slow down during provisioning and hold onto pods for a bit when we know some capacity is going to be freed by vertical scale-ins in the near future. You still have some coordination problems with consolidation there (it would be sad if you held a pod for a node that was going to get some capacity back and then consolidation got the node before you could place the pod there). This also directly trades-off of pod scaling responsiveness for packing efficiency and involves a whole slew of configuration to drive the behavior you want.
> 
> This leads to my final question, what is the behavior you are looking for here? How do we keep the consolidation process from causing to much disruption and cancelling out the potential savings? What config setups do you imagine doing this in? Are we only talking consolidateWhenEmpty or DoNotConsolidate or static nodepools here (you can probably use this feature with those nodepool settings today!)? Are you envisioning this working with ConsolidateWhenEmptyOrUnderutilized? How do you envision this working, can you walk through some scenarios of how this should work without causing feedback loops or excessive disruption?
> 
> Thanks! The team is busy and can always use help and contributions. I'm happy to work through this and help get it over the line once I understand more concretely what desired behavior is and have some ideas about how we can ensure it doesn't interact in unsafe ways with consolidation/provisioning.



At idealo we run ~100+ Spring Boot (JVM) microservices on `EKS with Karpenter`, and we're blocked on adopting `in-place` pod resize for exactly the failure modes described here.

The core issue: once the resize `subresource` mutates a pod's `resources.requests`, Karpenter's capacity model is stale. That leads to two concrete problems:

  - Over-scheduling: a pod resizes up, but Karpenter still thinks the node has the original headroom it schedules new pods onto a node that's already fully claimed
  - Consolidation thrash: a pod resizes down, Karpenter sees "free" capacity and consolidates, evicting pods that then resize back up on a new node repeat indefinitely

  For JVM workloads specifically this matters a lot. We currently use VPA with `updateMode: Auto` for right-sizing heap memory (we size JVM heap as a percentage of container memory limits). It works, but every VPA recommendation triggers a pod restart. Switching to `InPlaceOrRecreate` would eliminate restarts for CPU adjustments and upward memory resizes meaningfully reducing disruption for Kafka consumers and services with slow startup times.

We'd be happy to help test or give feedback on a design proposal if one is in progress.

**Comment by nathangeology** [comment]
Sounds good, I started working on something, but I don't like it very much yet and I hadn't thought too much about handling the scaling up side where you block provisioning due to the simulation thinking there is space that isn't actually there anymore. Sounds like we need a vertical pod scale down test case to check and measure 'consolidation thrash' and one where we scale some pods up vertically and check that we can still effectively provision new nodes appropriately afterwards. Anyone else have some other examples/test cases to look at here? Happy to collaborate on a design/proposal Ajay once I've got something that might be workable idea-wise for this. There's a lot of interesting data-model update/handling problems to solve here and then there's the disruption/provisioning coordination and timing problems and how we can configure/hint ways of dealing with them to the controllers. The biggest issue with the JVM style vertical scaling down case is that the best packing strategy for new nodes on a batch of incoming pods would be to wait a bit on some fraction of the pending pods so that things sort of cascade onto the cluster instead of hitting all at once and that can be dicey if the pods are needed urgently. This is less of a concern if there's a steady supply of new pods coming into to fill in the spaces as they open, but how do you guarantee that all the time while you are doing this? Dialing back consolidation and making it less aggressive could help with the thrash, but you'd be trading some on-going inefficiency for not messing with your now happily running JVM pods. If you have a pool of 'cheap' to disrupt pods, maybe we could use pod disruption cost to steer disruption towards those pods and away from the more expensive to disrupt JVM pods? The ugly thing is, this is all very situational and some requires a lot of depth of knowledge and willingness to experiment with configs by whomever is operating the cluster. Solving this in a more general, safe way is likely to be very conservative and leave a lot on the table efficiency-wise.

**Comment by msvticket** [comment]
This issue is not completed. While a [design](https://github.com/kubernetes-sigs/karpenter/blob/main/designs/in-place-pod-vertical-scaling.md) now apparently has been accepted it isn't actually fully implemented. It seems like the specific case of VPA being the controller that does the in place scaling would be solved by the merge of #3134, but even with that merged I would be hesitant to call this issue completed.

**Comment by jigisha620** [comment]
@msvticket ,
Curious what's driving the hesitation, is it a concern that the current design is scoped to VPA, or is there a different pod autoscaler you'd want to see supported?

**Comment by msvticket** [comment]
> [@msvticket](https://github.com/msvticket) , Curious what's driving the hesitation, is it a concern that the current design is scoped to VPA, or is there a different pod autoscaler you'd want to see supported?

When a solution that isn't bound to a specific scaler has been implemented I think this issue is completed. Hopefully kubernetes/autoscaler#9898 leads to a good design.

**Comment by DerekFrank** [comment]
should probably still be open until the implementation for https://github.com/kubernetes-sigs/karpenter/blob/main/designs/in-place-pod-vertical-scaling.md is in. We can close this issue after that, and open a new issue that references this and is a placeholder for https://github.com/kubernetes/autoscaler/issues/9898 support. We want folks interested in the broader discussions to be able to find it easily
